cpp/src/streaming/core/fanout.cpp

```diff
@@ -0,0 +1,441 @@
+ *
+ * @note Bounded fanout requires all the output channels to consume messages before
+ * the next message is sent/consumed from the input channel.
+ *
+ * @param ctx The context to use.
+ * @param ch_in The input channel to receive messages from.
+ * @param chs_out The output channels to send messages to.
+ * @return A node representing the bounded fanout operation.
+ */
+Node bounded_fanout(
+    std::shared_ptr<Context> ctx,
+    std::shared_ptr<Channel> ch_in,
+    std::vector<std::shared_ptr<Channel>> chs_out
+) {
+    ShutdownAtExit c1{ch_in};
+    ShutdownAtExit c2{chs_out};
+    auto& logger = ctx->logger();
+    co_await ctx->executor()->schedule();
+    while (true) {
+        auto msg = co_await ch_in->receive();
+        if (msg.empty()) {
+            break;
+        }
+
+        co_await send_to_channels(ctx.get(), std::move(msg), chs_out);
+        logger.trace("Sent message ", msg.sequence_number());
+    }
+
+    std::vector<Node> drain_tasks;
+    drain_tasks.reserve(chs_out.size());
+    for (auto& ch : chs_out) {
+        drain_tasks.emplace_back(ch->drain(ctx->executor()));
+    }
+    coro_results(co_await coro::when_all(std::move(drain_tasks)));
+    logger.trace("Completed bounded fanout");
+}
+
+/**
+ * @brief Unbounded fanout implementation.
+ *
+ * The implementation follows a pull-based model, where the send tasks request data from
+ * the recv task. There is one recv task that receives messages from the input channel,
+ * and there are N send tasks that send messages to the output channels.
+ *
+ * Main task operation:
+ * - There is a shared deque of cached messages, and a vector that indicates the next
+ * index of the message to be sent to each output channel.
+ * - Recv task awaits until the number of cached messages is equal to the latest sent
+ * message index by any of the send tasks. This notifies the recv task to pull a message
+ * from the input channel, cache it, and notify the send tasks about the new messages.
+ * recv task continues this process until the input channel is fully consumed.
+ * - Each send task awaits until there are more cached messages to send. Once notified, it
+ * determines the current end of the cached messages, and sends messages in the range
+ * [next_idx, end_idx). Once these messages have been sent, it updates the next index to
+ * end_idx and notifies the recv task.
+ *
+ * Additional considerations:
+ * - In the recv task loop, it also identifies the last completed message index by all
+ * send tasks. Message upto this index are no longer needed, and are purged from the
+ * cached messages.
+ * - When a send task fails to send a message, this means the channel may have been
+ * prematurely shut down. In this case, it sets its index to InvalidIdx. Recv task will
+ * filter out channels with InvalidIdx.
+ * - There two RAII helpers to ensure that the notification mechanisms are properly
+ * cleaned up when the unbounded fanout state goes out of scope/ encounters an error.
+ *
```

Comment:
Please do some work on this explanation. It intermingles low-level implementation with algorithmic stuff. The latter is more important.